{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "cd04f513",
   "metadata": {},
   "source": [
    "# ОБЛАСТЬ ВИДИМОСТИ ФУНКЦИЙ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4286485a",
   "metadata": {},
   "source": [
    "Создадим функцию print_root(), которая будет печатать корень степени n (по умолчанию n = 2, то есть вычисляется квадратный корень) из числа value. Внутри себя она будет содержать другую функцию root(), которая будет вычислять корень степени n из числа value.\n",
    "\n",
    "Напоминм, корень степени числа n из числа - это числа в степени 1/n."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "65fb32c6",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Зададим внешнюю функцию\n",
    "def print_root(value, n = 2):\n",
    "    #Зададим внутреннюю функцию, она будет являться вспомогательной\n",
    "    def root(value, n = 2):\n",
    "        result = value ** (1/n)\n",
    "        return result\n",
    "    #Получим результат из внутренней функции\n",
    "    res = root(value, n)\n",
    "    #Печатаем результат и не возвращаем его\n",
    "    print('Root of power', n, 'from', value, 'equals', res)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "65250e11",
   "metadata": {},
   "source": [
    "Вызовем функцию print_root с параметрами value = 81, n = 4, то есть вычислим корень четвертой степени из числа 81. Из школьного курса математики нам известно, что это будет 3, так как число 3 в четвертой степени дает 81."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0a57636d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Root of power 4 from 81 equals 3.0\n"
     ]
    }
   ],
   "source": [
    "print_root(81, 4)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "81326323",
   "metadata": {},
   "source": [
    "#### ***Что произошло в коде выше?***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7849519",
   "metadata": {},
   "source": [
    "Мы вызвали функцию print_root(), передав в нее параметры value = 81 и n = 4. Внутри себя функция print_root() вызвала внутреннюю функцию root()  с теми же параметрами для расчета корня. Внутри себя функция root() произвела расчет выражения *81 ** (1/4)*. Результат был занесен в переменную result и возвращен из функции с помощью return.\n",
    "\n",
    "Результат работы root() был занесен в переменную res, и ее значение было выведено на экран вместе с аргументами.\n",
    "\n",
    "Попробуем обратиться непосредственно к функции root:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "74d0fd5c",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'root' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[1;32mIn [3]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mroot\u001b[49m(\u001b[38;5;241m81\u001b[39m, \u001b[38;5;241m4\u001b[39m))\n",
      "\u001b[1;31mNameError\u001b[0m: name 'root' is not defined"
     ]
    }
   ],
   "source": [
    "print(root(81, 4))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3a9e89da",
   "metadata": {},
   "source": [
    "Эта ошибка переводится так: \"Ошибка имени: имя 'root' не определено\"."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8d9f3baf",
   "metadata": {},
   "source": [
    "Дело в том, что функция print_root скрывает, инкапсулирует, защищает (**инкапсулирует**) доступ ко внутренним функциям из основного скрипта, поэтому интерпретатор не может найти эту функцию, а функция root остается только для внутреннего использования."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f92c2142",
   "metadata": {},
   "source": [
    "Более того, давнное ограничение относится не только ко внутренней функции root(), но и ко всем переменны, объявленным внутри функции pritn_root()."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "07b6be80",
   "metadata": {},
   "source": [
    "Например, если попробуем обратиться к значению переменной res, которая была объявлена внутри функции print_root(), мы также получим ошибку:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "78ceea82",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'res' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[1;32mIn [4]\u001b[0m, in \u001b[0;36m<cell line: 1>\u001b[1;34m()\u001b[0m\n\u001b[1;32m----> 1\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mres\u001b[49m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'res' is not defined"
     ]
    }
   ],
   "source": [
    "print(res)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f0adf6d",
   "metadata": {},
   "source": [
    " В *Python* существует \"право на владение объектом\" - функцией/переменной. **Если функция/переменная находится в теле некоторой функции, обратиться к ней можно не из каждой точки программы.** Чтобы более детально разобраться в этом вопросе, нужно привести классификацию переменных и познакомиться с важным термином - ***разрешение переменных***."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f0b6b93",
   "metadata": {},
   "source": [
    "# РАЗРЕШЕНИЕ ПЕРЕМЕННЫХ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7270c84c",
   "metadata": {},
   "source": [
    "Мы занимаемся регистрацией данных о сотрудниках нашей компании в базе. Внесение в базу будем имитировать выводом информации на экран. Поля для ввода - имя и фамилия сотрудника. На экран должны быть выведены полное имя сотрудника (имя и фамилия) и название компании."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b9099be4",
   "metadata": {},
   "source": [
    "Создадим функцию register_employee() с аргументами name и surname (имя и фамилия сотрудника). Функция ничего не возвращает - просто выводит на экран фразу 'Employee {} is registered with the company {}', где вместо {} подставляется полное имя сотрудника (фамилия и имя) и название компании. Название компании хранится в переменной company_name, которая объявлена в основной части программы."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0034488",
   "metadata": {},
   "source": [
    "Внутри функции register_employee() объявим другую функцию - create_full_name(). Она возвращает полное имя сотрудника - результат сложения (*конкатенации*) строк с именем и фамилией:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "d6027e0c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Employee John Doe is registered with the company TheBlindMice\n"
     ]
    }
   ],
   "source": [
    "#объявляем внешнюю функцию для регистрации сотрудников\n",
    "def register_employee(name, surname):\n",
    "    #объявляем функцию для промежуточных вычислений\n",
    "    def create_full_name():\n",
    "        #функция использует внешние переменные name И surname\n",
    "        sep = ' ' #разделитель между именем и фамилией (пробел)\n",
    "        result = name + sep + surname #вычисляем полное имя\n",
    "        return result\n",
    "    full_name = create_full_name() #вызываем внутреннюю функцию\n",
    "    #выводим результат на экран, используя внешнюю переменную company_name\n",
    "    print(f'Employee {full_name} is registered with the company {company_name}')\n",
    "    \n",
    "company_name = 'TheBlindMice' #Название компании\n",
    "register_employee('John', 'Doe') #вызов функции"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1044c75c",
   "metadata": {},
   "source": [
    "Процесс поиска интерпретатором объекта, который скрывается за названием переменной, называется **разрешением**.\n",
    "\n",
    "Например, в функции register_employee() мы обращаемся к переменной company_name, однако данная переменная была объявлена за пределами функции. То, как интерпретатор понимает, что за переменной company_name скрывается строка 'TheBlindMice', и называется **разрешением**."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1f95f7ef",
   "metadata": {},
   "source": [
    "В питоне существует **четыре типа переменных в зависимости от их видимости**. Их разрешение будет идти от пункта 1 до пункта 4. Для любого разработчика очень важно понимать различия между ними, чтобы не сделать множество ошибок в попытках использовать переменные, которые использовать нельзя.\n",
    "\n",
    "В нашем учебном примере о регистрации сотрудников мы специально использовали все типы переменных. Давайте разбираться, что есть что:\n",
    "\n",
    "1. **Локальные переменные (local) (Локальная область видимости)** - это переменные, которые были объявлены в функции и используются непосредственно в ней. В разряд локальных переменных также входят аргументы функции.\n",
    "    - Аргументы функции register_employee(), name и surname, являются локальными переменными по отношению к этой функции.\n",
    "    - Функция create_full_name() является локальной для функции register_employee(). Попытка вызвать ее из основной части программы приведет к ошибке, которую мы уже видели раньше.\n",
    "    - Переменные sep и result являются локальными по отношению к функции create_full_name(). Они являются невидимыми для функции register_employee() и для основной части программы.\n",
    "    - Переменная full_name, в окторую заносится результат работы функции create_full_name(), является локальной переменной функции register_employee().\n",
    "\n",
    "2. **Нелокальные переменные (nonlocal) (Нелокальная область видимости/локальная область внешних функций)** - это переменные, которые были объявлены во внешней функции относительно рассматриваемой функции.\n",
    "    - Переменные name и surname являются нелокальными по отношению к функции create_full_name(). Они объявлены во внешней функции и используются во внутренней.\n",
    "    \n",
    "3. **Глобальные переменные(global) (Глобальная область видимости)** - это переменные, которые были объявлены непосредственно в основном блоке программы (вне функций).\n",
    "    - Переменная company_name объявлена в основной части программы (вне функции) и является глобальной. Обратите внимание, что она задана после объявления функции. Ошибки не возникает, потому что код выполняется построчно, а значит сначала создается переменная company_name(), а затем она уже используется в вызове функции register_employee(). Если бы поменяли вызов функции и объявление переменной местами, получили бы ошибку.\n",
    "    \n",
    "4. **Встроенные переменные (built-in) (Встроенная область видимости)** - это переменные и объекты, которые встроены в функционал питона изначально. Например, к ним относятся функции print, len, структуры данных list, dict, tuple И другие. В большинстве IDE *(интегрированная среда разработки, ИСР (integrated development environment), также единая среда разработки ЕСР - комплекс программных средств, используемых программистами для разработки программного обеспечения)*, таких как PyCharm, VS Code и Jupyter, имена таких переменных подсвечиваются специальным цветом.\n",
    "    - Функция print() является встроенной функцией питона. К ней можно обращаться из любой части программы."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f46c36f5",
   "metadata": {},
   "source": [
    "Когда интерпретатор встречает в коде функции ссылку (переменную) на какой-то объект, он начинает искать его среди **локальных** переменных, затем переключается на **нелокальные**, потом на **глобальные** и, наконец, ищет переменную среди **встроенных** оъектов. Если поиск оказался безрезультатным, возникнет ошибка."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "138e0d26",
   "metadata": {},
   "source": [
    "Разберемся, как работает разрешение (как интерпретатор находит значения переменных) в нашем примере:\n",
    "\n",
    "1. В вызове функции create_full_name() мы используем переменные name и surname. Интерпретитор сначала пытается нати их среди **локальных** переменных самой функции, но попытка оказывается безуспешной. ТОгда проводится проверка на совпадение имен среди **нелокальных** переменных, и она оказывается успешной - перменные объявлены во внешней функции.\n",
    "\n",
    "2. В вызове функции register_employee() мы используем переменную company_name. Интерпретатор сначала (безуспешно) ищет ее среди локальных переменных. Поиск среди нелокальных переменных не имеет смысла, так как функция является внешней. Тогда переходим к глобальным переменным. И - о, чудо! Переменная company_name была объявлена до вызова функции, а значит мы смогли найти ее в глобальной области видимости.\n",
    "\n",
    "3. В той же функции register_employee() мы используем функцию print(). Но интерпретатор не знает, что эта функция - встроенная. Значит, проводится поиск среди локальных имен, пропуская нелокальный поиск, потом ведется поиск среди глобальных переменных, и только на слое встроенных имен поиск оказывается успешным."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "88bda1fb",
   "metadata": {},
   "source": [
    "Рассмотрим различия между типами переменных, некорректно обращаясь к ним.\n",
    "\n",
    "#### Например, что будет, если мы попробуем обратиться к переменной result во внешней функции."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "b13424da",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'result' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[1;32mIn [8]\u001b[0m, in \u001b[0;36m<cell line: 19>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     17\u001b[0m company_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTheBlindMice\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m     18\u001b[0m \u001b[38;5;66;03m#вызов функции\u001b[39;00m\n\u001b[1;32m---> 19\u001b[0m \u001b[43mregister_employee\u001b[49m\u001b[43m(\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mJohn\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m,\u001b[49m\u001b[43m \u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43mDoe\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[1;32mIn [8]\u001b[0m, in \u001b[0;36mregister_employee\u001b[1;34m(name, surname)\u001b[0m\n\u001b[0;32m     11\u001b[0m full_name \u001b[38;5;241m=\u001b[39m create_full_name()\n\u001b[0;32m     12\u001b[0m \u001b[38;5;66;03m#пытаемся обратиться к result из основной части программы\u001b[39;00m\n\u001b[1;32m---> 13\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mresult\u001b[49m)\n\u001b[0;32m     14\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;124mf\u001b[39m\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mEmployee \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mfull_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m is registered with the company \u001b[39m\u001b[38;5;132;01m{\u001b[39;00mcompany_name\u001b[38;5;132;01m}\u001b[39;00m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'result' is not defined"
     ]
    }
   ],
   "source": [
    "def register_employee(name, surname):\n",
    "    #name и surname - Локальные переменные register_employee\n",
    "    def create_full_name():\n",
    "        #sep - локальная переменная для create_full_name\n",
    "        sep = ' '\n",
    "        #name, surname - нелокальные переменные для create_full_name\n",
    "        #result - локальная переменная для create_full_name\n",
    "        result = name + sep + surname\n",
    "        return result\n",
    "    #full_name - локальная перменная register_employee\n",
    "    full_name = create_full_name()\n",
    "    #пытаемся обратиться к result из основной части программы\n",
    "    print(result)\n",
    "    print(f'Employee {full_name} is registered with the company {company_name}')\n",
    "    \n",
    "#company_name - глобальная переменная\n",
    "company_name = 'TheBlindMice'\n",
    "#вызов функции\n",
    "register_employee('John', 'Doe')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "12c403f4",
   "metadata": {},
   "source": [
    "Мы получили ошибку, которая говорит о том, что переменная result не объявлена в функции register_employee(). Так произошло, потому что функция register_employee() ничего не знает о переменной result, так как она является локальной для функции create_full_name(). На языке нашей схемы: мы попытались пригласить на второй этаж жителя с третьего этажа, что противоречит правилам."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b785087d",
   "metadata": {},
   "source": [
    "Такую же ошибку мы получим, если попробуем обратиться к переменной full_name из основной части программы:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "d6c6f9d6",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Employee <function register_employee.<locals>.create_full_name at 0x00000229255820D0> is registered with the company TheBlindMice\n"
     ]
    },
    {
     "ename": "NameError",
     "evalue": "name 'full_name' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[1;32mIn [9]\u001b[0m, in \u001b[0;36m<cell line: 11>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      9\u001b[0m company_name \u001b[38;5;241m=\u001b[39m \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mTheBlindMice\u001b[39m\u001b[38;5;124m'\u001b[39m\n\u001b[0;32m     10\u001b[0m register_employee(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mJone\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mDoe\u001b[39m\u001b[38;5;124m'\u001b[39m)\n\u001b[1;32m---> 11\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mfull_name\u001b[49m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'full_name' is not defined"
     ]
    }
   ],
   "source": [
    "def register_employee(name, surname):\n",
    "    def create_full_name():\n",
    "        sep = ' '\n",
    "        result = name + sep + surname\n",
    "        return result\n",
    "    full_name = create_full_name\n",
    "    print(f'Employee {full_name} is registered with the company {company_name}')\n",
    "    \n",
    "company_name = 'TheBlindMice'\n",
    "register_employee('Jone', 'Doe')\n",
    "print(full_name) #попытаеся обратиться к full_name из основной части программы"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6c7f9c30",
   "metadata": {},
   "source": [
    "Так как full_name - локальная переменная для функции register_full_name(), и основная часть программы о ней ничего не знает. На языке нашей схемы: мы пытаемся пригласить на первый этаж жителя со второго этажа, что является невозможным."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a4e4cbc5",
   "metadata": {},
   "source": [
    "Есть еще одно **очень важное ограничение:** если в коде функции происходит переопределение глобальной или нелокальной переменной, то необходимо показать, что та или иная переменная является глобальной или нелокальной.\n",
    "\n",
    "Например, если внутри тела функции происходит увеличение/уменьшение глобальной переменной с помощью операторов +=/-=, то такое изменение значения глобальной переменной вызовет ошибку.\n",
    "\n",
    "Если руководствоваться аналогией с нашей многоэтажкой: приглашая жителя снизу, мы не можем переодеть его, явно не спросив об этом."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9066691c",
   "metadata": {},
   "source": [
    "Рассмотрим это ограничение на примере.\n",
    "\n",
    "Пусть у нас есть переменная global_counter, в которой мы храним количество, например, количество товаров на складе. КОгда товар поступает на склад, мы увеличиваем его количество на 1.\n",
    "\n",
    "Для этого напишем функцию add_one(), которая увеличивает значение глобальной переменной global_counter на 1, вызовем ее и попробуем посмотреть на значение переменной.\n",
    "\n",
    "Напишем функцию, которая увеличивает значение глобальной перменной на 1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "d199cf20",
   "metadata": {},
   "outputs": [
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'global_counter' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[1;32mIn [10]\u001b[0m, in \u001b[0;36m<cell line: 8>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      5\u001b[0m     global_counter \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m      7\u001b[0m \u001b[38;5;66;03m#Запустим функцию add_one\u001b[39;00m\n\u001b[1;32m----> 8\u001b[0m \u001b[43madd_one\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n\u001b[0;32m      9\u001b[0m \u001b[38;5;66;03m#Напечатаем значение переменной global_counter\u001b[39;00m\n\u001b[0;32m     10\u001b[0m \u001b[38;5;28mprint\u001b[39m(global_counter)\n",
      "Input \u001b[1;32mIn [10]\u001b[0m, in \u001b[0;36madd_one\u001b[1;34m()\u001b[0m\n\u001b[0;32m      4\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21madd_one\u001b[39m():\n\u001b[1;32m----> 5\u001b[0m     global_counter \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n",
      "\u001b[1;31mUnboundLocalError\u001b[0m: local variable 'global_counter' referenced before assignment"
     ]
    }
   ],
   "source": [
    "#Создадим глобальную переменную и приравняем ее нулю\n",
    "global_counter = 0\n",
    "#Создадим функцию, которая прибавляет 1 к переменной global_counter\n",
    "def add_one():\n",
    "    global_counter += 1\n",
    "    \n",
    "#Запустим функцию add_one\n",
    "add_one()\n",
    "#Напечатаем значение переменной global_counter\n",
    "print(global_counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5097578d",
   "metadata": {},
   "source": [
    "Перевод ошибки: *«К локальной переменной 'global_counter' обратились до объявления переменной в коде»*. Это означает, что интерпретатор считает эту переменную локальной, хотя в основной части программы есть одноимённая переменная."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3ac0897c",
   "metadata": {},
   "source": [
    "Но в чём проблема? В предыдущем примере мы спокойно использовали значение глобальной переменной company_name в теле функции, а сейчас возникает ошибка."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3fe28813",
   "metadata": {},
   "source": [
    "Ответ следующий: это специальный механизм защиты глобальных переменных в Python. Когда мы пытаемся перезаписать (изменить) значение глобальной переменной, она является для нас недоступной."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1216ee32",
   "metadata": {},
   "source": [
    "Как именно возникает такая ошибка?\n",
    "\n",
    "Присмотримся повнимательнее к операции присваивания. Слева от оператора присваивания (=) стоит переменная global_counter. Когда выполняется операция присваивания, в памяти выделяется место для новой локальной переменной global_counter. Однако её значение ещё не было задано — было выделено только место для хранения, но само содержимое «не привезли».\n",
    "\n",
    "Эта же переменная стоит и справа от оператора присваивания(=): global_counter = global_counter + 1. Когда мы обращаемся к переменной global_counter справа, то global_counter считается в этот момент уже локальной, а не глобальной (и, напомним, её значение ещё не было задано). Возникает противоречие: мы пытаемся обратиться к переменной, значение которой ещё не было определено. Отсюда и возникает ошибка UnboundLocalError."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ed14539b",
   "metadata": {},
   "source": [
    "**Важное замечание:** стандарты разработки на Python не рекомендуют изменять в функциях глобальные переменные, однако такая возможность предусмотрена. Для этого необходимо добавить оператор global внутри функции перед той переменной, которую вы хотите изменить глобально."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "970cda8d",
   "metadata": {},
   "source": [
    "В данном случае, чтобы изменить переменную на глобальном уровне, нам необходимо добавить строку global global_counter:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "4e4fcd0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "global_counter = 0\n",
    "\n",
    "def add_one():\n",
    "    #Обозначим, что переменная global_counter является глобальной\n",
    "    global global_counter\n",
    "    global_counter += 1\n",
    "    \n",
    "add_one()\n",
    "print(global_counter)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c442d234",
   "metadata": {},
   "source": [
    "Теперь функция сработала без ошибок, а переменная global_counter увеличилась на 1 в глобальном смысле.\n",
    "\n",
    "Аналогичного поведнеия следует ожидать от вложенных функций и нелокальных переменных.\n",
    "\n",
    "Например, создадим функцию inner_function(), кооторая использует переменную enclosing_counter, созданную во внешней функции outer_function(). Она также должна увеличивать значение этой переменной на 1 и сразу напечатать ее значение. Все, как в прошлой задаче, но теперь переменная является нелокальной."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "af01c39f",
   "metadata": {},
   "outputs": [
    {
     "ename": "UnboundLocalError",
     "evalue": "local variable 'enclosing_counter' referenced before assignment",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mUnboundLocalError\u001b[0m                         Traceback (most recent call last)",
      "Input \u001b[1;32mIn [1]\u001b[0m, in \u001b[0;36m<cell line: 15>\u001b[1;34m()\u001b[0m\n\u001b[0;32m     12\u001b[0m     inner_function()\n\u001b[0;32m     14\u001b[0m \u001b[38;5;66;03m#запустим внешнюю фнукцию\u001b[39;00m\n\u001b[1;32m---> 15\u001b[0m \u001b[43mouter_function\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[1;32mIn [1]\u001b[0m, in \u001b[0;36mouter_function\u001b[1;34m()\u001b[0m\n\u001b[0;32m     10\u001b[0m     \u001b[38;5;28mprint\u001b[39m(enclosing_counter)\n\u001b[0;32m     11\u001b[0m \u001b[38;5;66;03m#запустим внутреннюю функцию из внешней\u001b[39;00m\n\u001b[1;32m---> 12\u001b[0m \u001b[43minner_function\u001b[49m\u001b[43m(\u001b[49m\u001b[43m)\u001b[49m\n",
      "Input \u001b[1;32mIn [1]\u001b[0m, in \u001b[0;36mouter_function.<locals>.inner_function\u001b[1;34m()\u001b[0m\n\u001b[0;32m      6\u001b[0m \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21minner_function\u001b[39m():\n\u001b[0;32m      7\u001b[0m     \u001b[38;5;66;03m#прибавим 1 к enclosing_counter\u001b[39;00m\n\u001b[1;32m----> 8\u001b[0m     enclosing_counter \u001b[38;5;241m+\u001b[39m\u001b[38;5;241m=\u001b[39m \u001b[38;5;241m1\u001b[39m\n\u001b[0;32m      9\u001b[0m     \u001b[38;5;66;03m#Напечатаем значение enclosing_counter\u001b[39;00m\n\u001b[0;32m     10\u001b[0m     \u001b[38;5;28mprint\u001b[39m(enclosing_counter)\n",
      "\u001b[1;31mUnboundLocalError\u001b[0m: local variable 'enclosing_counter' referenced before assignment"
     ]
    }
   ],
   "source": [
    "#Внешняя функция\n",
    "def outer_function():\n",
    "    #Создадим переменную, относящуюся к внешней функции\n",
    "    enclosing_counter = 0\n",
    "    #Внутренняя функция\n",
    "    def inner_function():\n",
    "        #прибавим 1 к enclosing_counter\n",
    "        enclosing_counter += 1\n",
    "        #Напечатаем значение enclosing_counter\n",
    "        print(enclosing_counter)\n",
    "    #запустим внутреннюю функцию из внешней\n",
    "    inner_function()\n",
    "    \n",
    "#запустим внешнюю фнукцию\n",
    "outer_function()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e60430e9",
   "metadata": {},
   "source": [
    "Как и ожидалось, мы получили ошибку. Причём ошибка того же характера: интерпретатор полагает, что enclosing_counter — это локальная переменная, которая ещё не была объявлена."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "1983e57c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n"
     ]
    }
   ],
   "source": [
    "def outer_function():\n",
    "    enclosing_counter = 0\n",
    "    def inner_function():\n",
    "        #с помощью оператора nonlocal покажем, что переменная enclosing_counter\n",
    "        #находится во внешней функции\n",
    "        nonlocal enclosing_counter\n",
    "        enclosing_counter += 1\n",
    "        print(enclosing_counter)\n",
    "    inner_function()\n",
    "    \n",
    "outer_function()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ae5e1ff5",
   "metadata": {},
   "source": [
    "Теперь ошибка не возникает, а поведение функции оказывается совершенно предсказуемым.\n",
    "\n",
    "У операции, которую мы только что совершили, есть свое название. В данном случае функция inner_function является **функцией-замыканием** - она использует в своем коде ссылки на переменные, которые были объявлены во внешней функции, но не в основном коде программы."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34ea185b",
   "metadata": {},
   "source": [
    "Возможно ли изменять встроенные объекты? Например, можно ли присваивать значение переменной с названием len, хотя len изначально используется в Python для подсчёта числа элементов в структуре данных?\n",
    "\n",
    "Да, возможно. Однако это крайне нежелательно, поскольку теряется доступ к возможностям исходного объекта или функции."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5b13ea0",
   "metadata": {},
   "source": [
    "### Задание 2.3\n",
    "\n",
    "Что будет выведено в результате выполнения следующего кода?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "87b8568c",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'printer' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[1;32mIn [4]\u001b[0m, in \u001b[0;36m<cell line: 5>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      2\u001b[0m     \u001b[38;5;28;01mdef\u001b[39;00m \u001b[38;5;21mprinter\u001b[39m():\n\u001b[0;32m      3\u001b[0m         \u001b[38;5;28mprint\u001b[39m(msg)\n\u001b[1;32m----> 5\u001b[0m \u001b[43mprinter\u001b[49m(\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mHello\u001b[39m\u001b[38;5;124m'\u001b[39m)\n",
      "\u001b[1;31mNameError\u001b[0m: name 'printer' is not defined"
     ]
    }
   ],
   "source": [
    "def print_msg(msg):\n",
    "    def printer():\n",
    "        print(msg)\n",
    "        \n",
    "printer('Hello')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "4c8dea8d",
   "metadata": {},
   "source": [
    "### Задание 2.3\n",
    "\n",
    "Вашему коллеге было дано задание написать во внешней функции total_multiplier внутреннюю функцию-замыкание multiplier, которая увеличивает значение переменной total на значение 10*n, где n - аргумент внешней функции.\n",
    "\n",
    "Он написал следующий код, однако при вызове функции total_multiplier(10) возникает ошибка \"UnboundLocalError: local variable 'total' referenced before assignment."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "4a53a2d9",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "100"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "def total_multiplier(n):\n",
    "    total = 0\n",
    "    def multiplier():\n",
    "        nonlocal total\n",
    "        product = 10 * n\n",
    "        total += product\n",
    "        return total\n",
    "    return multiplier()\n",
    "\n",
    "total_multiplier(10)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c753134c",
   "metadata": {},
   "source": [
    "# ПРИМЕНЕНИЕ ОБЛАСТИ ВИДИМОСТИ ПЕРЕМЕННЫХ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5cd92cf3",
   "metadata": {},
   "source": [
    " В Python можно возвращать не только результаты вычислений, но и другие функции. Например, можно возвращать внутренние функции. Страшно? Давайте разбираться.\n",
    " \n",
    "Давайте рассмотрим небольшой пример такого механизма.\n",
    "\n",
    "Пусть у нас есть внешняя функция outer(), в которой объявлена перменная х и внутренняя функция inner(), которая просто выводит значение перменной х на экран. Но отличие данной конструкции от тех, тчо мы рассматривали ранее, состоит в том, что внешняя функция возвращает внутреннюю."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "515ade43",
   "metadata": {},
   "outputs": [],
   "source": [
    "#объявляем внешнюю фнукцию\n",
    "def outer():\n",
    "    x = 1 # создаем локальную перменную\n",
    "    #объявляем внутрненнюю функцию\n",
    "    def inner():\n",
    "        print('x in outer function: ', x)\n",
    "    #внешняя функция возвращает внутреннюю\n",
    "    return inner"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8f6d50c4",
   "metadata": {},
   "source": [
    "Попробуем вызвать функцию outer() и положить результат ее работы в переменную Out. Затем выведем содержимое:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "06814eb1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<function outer.<locals>.inner at 0x000001C2FDE4FF70>\n"
     ]
    }
   ],
   "source": [
    "out = outer()\n",
    "print(out)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "285e3810",
   "metadata": {},
   "source": [
    "Итак, результат работы функции outer() — это функция. Попробуем вызвать и её:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "af30737d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "x in outer function:  1\n"
     ]
    }
   ],
   "source": [
    "out()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a6875da5",
   "metadata": {},
   "source": [
    "Таким образом, в переменной out содержится ссылка на функцию inner().\n",
    "\n",
    "Теперь посмотрим на более короткую запись того, как можно «достучаться» до функции inner(). Действуем по принципу матрёшки: открываем большую матрёшку → вызываем внешнюю функцию → получаем внутреннюю функцию, открываем матрёшку поменьше → вызываем внутреннюю функцию:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "3f1ce339",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "x in outer function:  1\n"
     ]
    }
   ],
   "source": [
    "outer()()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1913ba34",
   "metadata": {},
   "source": [
    "Теперь, обладая этим знанием, мы сможем справиться со следующим примером.\n",
    "\n",
    "Перед нами стоит следующая задача: подсчитать, сколько раз функция вызывается в программе. Такая задача может возникнуть, когда функция вызвается при каком-то определенном условии и вам необходимо отслеживать, сколько раз это условия выполняется. Например, у вас есть система рекомендаций музыки, и вы должны отследить, сколько раз с момента запуска программы вызвалась функция переключения музыки, чтобы понять, нравится ли пользователю представленная ему подборка."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "9b85e012",
   "metadata": {},
   "source": [
    "Давайте в теле внешней функции counter() напишем внутреннюю функцию-счётчик add(), которая с каждым запуском будет считать, сколько раз её вызвали. \n",
    "\n",
    "Во внешней функции объявим переменную number. Во внутренней функции будем увеличивать значение этой переменной на 1. Обязательно будем использовать оператор nonlocal, чтобы показать, что текущее значение счётчика необходимо брать из внешней функции и оно будет изменяться. Говоря профессиональным языком, мы создаём замыкание."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "bf27d553",
   "metadata": {},
   "outputs": [],
   "source": [
    "#функция, которая создает счетчик\n",
    "def counter():\n",
    "    #начальное значение счетчика - 0\n",
    "    number = 0\n",
    "    #функция add будет каждый раз прибавлять к счетчики 1 при запуске\n",
    "    def add():\n",
    "        #сообщаем, что number берем из внешней функции\n",
    "        nonlocal number\n",
    "        #увеличиваем значение счетчика на 1\n",
    "        number += 1\n",
    "        #возвращаем текущее число запусков счетчика\n",
    "        return number\n",
    "    #возвращаем не результат вычислений, а непосредственно саму функцию add\n",
    "    #без круглых скобок!\n",
    "    return add"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ddd6b41d",
   "metadata": {},
   "source": [
    "Теперь при вызове функции counter() будет возвращаться функция в виде объекта, который можно использовать отдельно от изначальной функции counter(). Эта функция, хранящаяся в виде нового объекта, будет обладать возможностями функции add(), которая была задана внутри внешней функции counter().\n",
    "\n",
    "Прежде чем воспользоваться новым счётчиком, его необходимо создать, вызвав функцию counter(), и сохранить в какую-нибудь переменную, например counter1:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "119b137e",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n"
     ]
    }
   ],
   "source": [
    "#в переменную counter1 сохраняем новый счетчик\n",
    "counter1 = counter()\n",
    "\n",
    "#проверим, что counter1 действительно является вызваемым объектом,\n",
    "#то есть функцией.Для этого воспользуемся встроенной функцией callable:\n",
    "print(callable(counter1))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c6e51216",
   "metadata": {},
   "source": [
    "Чтобы воспользоваться только что полученной функцией counter1 для увеличения хранящегося в ней значения на 1, достаточно написать круглые скобки после переменной counter1, то есть воспользоваться привычным синтаксисом вызова функции:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "123355c1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "2\n",
      "3\n"
     ]
    }
   ],
   "source": [
    "# Запустим counter как обычную функцию\n",
    "print(counter1())\n",
    "print(counter1())\n",
    "print(counter1())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba26993c",
   "metadata": {},
   "source": [
    "Как видите, функция counter1 действительно считает, сколько раз её запустили из кода. \n",
    "\n",
    "Давайте убедимся, что каждый счётчик считает только свои запуски, то есть не происходит глобального изменения переменной number во всей программе:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "f2f8b1d1",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Counter 1: 1\n",
      "Counter 1: 2\n",
      "Counter 2: 1\n",
      "Counter 1: 3\n",
      "Counter 2: 2\n",
      "Counter 2: 3\n"
     ]
    }
   ],
   "source": [
    "#Созадим два различных счетчика\n",
    "counter1 = counter()\n",
    "counter2 = counter()\n",
    "#Будем запускать врвзнобой разные счетчики\n",
    "\n",
    "print(\"Counter 1:\", counter1())\n",
    "print(\"Counter 1:\", counter1())\n",
    "print(\"Counter 2:\", counter2())\n",
    "print(\"Counter 1:\", counter1())\n",
    "print(\"Counter 2:\", counter2())\n",
    "print(\"Counter 2:\", counter2())"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d9bd3165",
   "metadata": {},
   "source": [
    "В нашем коде мы дважды вызвали функцию counter() и записали результаты её работы в две различные переменные: counter1 и counter2. В каждой из них содержится ссылка на внутреннюю функцию add(), в которой происходит увеличение нелокальной переменной number на 1. \n",
    "\n",
    "Ключевым в понимании кода выше является тот факт, что для каждого из вызовов функции counter() переменная number является локальной, то есть для каждого вызова она своя. Переменная number создаётся при каждом вызове внешней функции counter(), а мы вызвали её дважды.\n",
    "\n",
    "Таким образом, когда мы вызвали функцию counter() в первый раз, создалась первая переменная number:\n",
    "\n",
    "counter1 = counter()\n",
    "\n",
    "Когда мы вызвали функцию counter() во второй раз, создалась вторая переменная number:\n",
    "\n",
    "counter2 = counter()\n",
    "\n",
    "Однако она не имеет никакого отношения к первой. **Изменение первой переменной не влечёт за собой изменение второй, и наоборот.**\n",
    "\n",
    "Поэтому, когда мы вызываем функцию add() от имени переменных counter1 и counter2, мы изменяем значения двух различных переменных number. \n",
    "\n",
    "Получается, что для каждой функции ведётся индивидуальный подсчёт её вызовов.\n",
    "\n",
    "***В этом и заключается преимущество использования замыканий по сравнению с модификацией глобальных переменных в теле функции: мы увеличиваем значение переменной только в рамках одной функции.***"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b7e2bb0d",
   "metadata": {},
   "source": [
    "### Когда это может вам понадобиться?\n",
    "\n",
    "На самом деле замыкания — это инструмент, который позволяет избежать использования глобальных переменных, и обеспечивает некоторую форму сокрытия переменных: мы обеспечиваем работу двух одинаковых интерфейсов (функций) независимо друг от друга. К тому же замыкания — это первый шаг к пониманию более сложных конструкций, которые мы изучим далее — **декораторов.**\n",
    "\n",
    "Представим ситуацию ↓\n",
    "\n",
    "Для построения модели, которая будет предсказывать поведение пользователей в приложении (например, сделает клиент заказ или нет), нам может пригодиться информация о том, сколько раз каждый клиент кликал на кнопку «Купить». \n",
    "\n",
    "Предположим, у нас уже есть функция, которая привязана к этой кнопке, то есть она вызывается при нажатии кнопки покупки. Нам необходимо постоянно вести подсчёт, сколько раз вызывается эта функция, но не в глобальном смысле (не для всех пользователей в общем), а для каждого пользователя независимо от других. Тут нам и могут пригодиться замыкания. Мы можем организовать внутреннюю функцию-замыкание, в которой будем увеличивать количество нажатий кнопки «Купить»."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7fd6f631",
   "metadata": {},
   "source": [
    "### Задание 2.5\n",
    "\n",
    "Напишите функцию-копилку с названием saver(), которая не принимает никаких аргументов. Она должна возвращать внутреннюю функцию adder(), которая принимает на вход одно число и возвращает сумму в копилке после прибавления числа.\n",
    "\n",
    "Изначально в новой копилке хранится 0."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "0859e88d",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "25\n",
      "125\n",
      "144\n"
     ]
    }
   ],
   "source": [
    "def saver():\n",
    "    number = 0\n",
    "    def adder(x):\n",
    "        nonlocal number\n",
    "        number += x\n",
    "        return number\n",
    "    return adder\n",
    "\n",
    "pig = saver()\n",
    "print(pig(25))\n",
    "print(pig(100))\n",
    "print(pig(19))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "f3706dda",
   "metadata": {},
   "source": [
    "# РЕКУРСИЯ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "91b5737f",
   "metadata": {},
   "source": [
    "В Data Science и программировании вообще рекурсии — очень важный и полезный инструмент при разработке алгоритмов. Вот несколько примеров, где вы можете столкнуться с рекурсией:\n",
    "\n",
    "    * В алгоритме дерева решений — одном из главных алгоритмов машинного обучения, который мы будем изучать в нашем курсе. Рекурсия лежит в основе его реализации.\n",
    "    * В различных методах сортировки: сортировка вставками (Merge Sort), быстрая сортировка (Quick Sort).\n",
    "    * В методах динамического программирования, которое используется в обучении с подкреплением.\n",
    "    * При генерации изображений в видеоиграх."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0422e293",
   "metadata": {},
   "source": [
    "# ОПРЕДЕЛЕНИЕ РЕКУРСИИ И ЕЁ ПРИМЕНЕНИЕ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a89d5566",
   "metadata": {},
   "source": [
    "→ Вы наверняка замечали, что если поставить два зеркала друг напротив друга, отражения в них начнутся повторяться и всё больше и больше уходить вдаль. При этом получается так, что в зеркале появляется отражение его самого, которое всё продолжает появляться в уменьшенной копии. Это явление, когда какой-то объект содержит в себе самого себя, называется рекурсией.\n",
    "\n",
    "Рекурсии могут возникнуть и в бытовых ситуациях. Например, нам необходимо совершить одни и те же действия (переместить, удалить, переименовать, задать значки и т. д.) со всеми файлами во вложенных папках. Каждая папка содержит несколько папок, в этих папках также есть папки, в тех — ещё папки. Так будет продолжаться до тех пор, пока мы не дойдём до конечной папки, в которой не будет вложенных папок.\n",
    "\n",
    "Функции также могут быть рекурсиями. При этом подразумевается, что рекурсивная функция в своём теле использует ссылку на саму себя. Однако, в отличие от двух зеркал, в рекурсивной функции должно быть прописано такое условие, при выполнении которого функция перестаёт вызывать сама себя. Такое условие будем называть **условием выхода из рекурсии** (его также называют **базовым случаем**)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ba5ef8f7",
   "metadata": {},
   "source": [
    "Например, для прохода по вложенным папкам условие выхода из рекурсии — «в текущей папке больше нет вложенных папок» (в этом случае рекурсия останавливается).\n",
    "\n",
    "Чтобы научиться применять рекурсию, вам необходимо будет «расширить границы своего мышления», так как поначалу синтаксис может казаться вам странным. Однако рекурсивные функции позволяют быстро и лаконично реализовать свою идею в коде. Попробуем научиться «мыслить рекурсивно»."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "cdac7f1a",
   "metadata": {},
   "source": [
    "# СУММА ЭЛЕМЕНТОВ СПИСКА ЧЕРЕЗ РЕКУРСИЮ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "acec3577",
   "metadata": {},
   "source": [
    "Для начала рассмотрим простой, но важный для понимания пример. Задача, которую мы уже решали множество раз различными способами — найти сумму элементов в списке. На этот раз решим её с помощью рекурсии, чтобы понять, как она работает.\n",
    "\n",
    "У нас есть список чисел, например [10, 21, 24, 12]. Представим, что мы забыли, как складывать несколько чисел сразу, но помним, как складывать два числа между собой.\n",
    "\n",
    "У нас есть четыре числа.\n",
    "\n",
    "Что такое сумма чисел в списке [10, 21, 24, 12]? Это 10 + сумма в списке [21, 24, 12].\n",
    "\n",
    "Что такое сумма списка [21, 24, 12]?  Это 21 + сумма [24, 12].\n",
    "\n",
    "А что такое сумма списка [24, 12]? Это 24 + 12 = 36.\n",
    "\n",
    "Элементов в списке больше не осталось.\n",
    "\n",
    "А теперь идём в обратную сторону: 36 + 21 = 57; 57 + 10 = 67.\n",
    "\n",
    "10+21+24+12=10+(21+24+12)=10+(21+(24+12))=10+(21+36)=10+57=67\n",
    "\n",
    "Не напомнили ли вам наши вычисления пример с зеркалом? Мы несколько раз подряд производили одну и ту же операцию — складывали первое число со списком из оставшихся, но, в отличие от зеркала, когда мы дошли точки остановки (список закончился), мы прекратили выполнение операций и пошли в обратную сторону.\n",
    "\n",
    "Это и есть пример рекурсивных вычислений. У нас есть операция (функция), которая обращается к самой себе до тех пор, пока не выполнится условие.\n",
    "\n",
    "Посмотрим теперь, как это выглядит в коде. Объявим функцию sum_lst() с одним аргументом lst — списком, по которому нужно рассчитать сумму.\n",
    "\n",
    "Прежде всего в рекурсивной функции прописывается условие выхода из рекурсии — условие остановки. Это условие, когда функция возвращает определённое значение. Более детально о его важности мы поговорим чуть позже. Наше условие — список пуст (его длина равна 0). Если это условие выполнилось, мы возвращаем 0 (сумма пустого списка).\n",
    "\n",
    "В противном случае, функция sum_lst() возвращает сумму первого элемента в списке с результатом вызова самой себя, но в аргументы уже передаётся список, из которого исключён первый элемент:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "988350d8",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[10, 21, 24, 12]\n",
      "[21, 24, 12]\n",
      "[24, 12]\n",
      "[12]\n",
      "[]\n",
      "67\n"
     ]
    }
   ],
   "source": [
    "def sum_lst(lst):\n",
    "    #выводим текущее значение lst\n",
    "    print(lst)\n",
    "    #задаем условие выхода из рекурсии\n",
    "    if len(lst) == 0: return 0\n",
    "    #во всех других случаях возвращаем сумму первого элекмента списка\n",
    "    #и результат суммирования оставшихся\n",
    "    return lst[0] + sum_lst(lst[1:])\n",
    "\n",
    "my_lst = [10, 21, 24, 12]\n",
    "print(sum_lst(my_lst))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7f154980",
   "metadata": {},
   "source": [
    "    1. При первом вызове sum_lst() в основной части программы локальная переменная lst = [10, 21, 24, 12]. Длина lst не равна 0, а значит выполняется основной блок. В нём мы извлекаем из списка lst первый по счёту элемент (число 10) и складываем его с результатом вызова функции sum_lst(), но передаём в неё не весь список, а только оставшиеся числа: [21, 24, 12]. Так как функция ещё не получила окончательного ответа, то выполнение (возвращение значения оператором return) текущего вызова sum_lst() откладывается, пока не выполнится новый.\n",
    "\n",
    "    2.Функция sum_lst() выполняется вновь. Теперь локальная переменная lst = [21, 24, 12]. Длина lst не равна 0, а значит выполняется основной блок. В нём мы извлекаем из списка lst первый по счёту элемент (число 21) и складываем его с результатом вызова функции sum_lst(), но передаём в неё не весь список, а только оставшиеся числа: [24, 12]. Выполнение текущего вызова sum_lst() откладывается, пока не выполнится новый.\n",
    "\n",
    "\n",
    "    3. И снова происходит выполнение функции sum_lst(). Всё то же самое, что и раньше, только теперь lst = [24, 12]. Первый элемент списка — число 24, а оставшийся — [12]. Вызываем функцию для оставшихся элементов. Выполнение текущего вызова sum_lst() откладывается, пока не выполнится новый.\n",
    "\n",
    "\n",
    "    4. Мы вновь внутри функции sum_lst(), но теперь lst = [12]. Список не пустой, а значит выполнение функции продолжается. Первый элемент списка (он же единственный) — число 12. Оставшиеся элементы — пустой список. Вызываем функцию для него. Выполнение текущего вызова sum_lst() откладывается, пока не выполнится новый.\n",
    "\n",
    "\n",
    "    5. Финал (ну, почти)! Список lst = []. Его длина равна 0, а значит функция наконец-то что-то возвращает, а именно — число 0.\n",
    "\n",
    "\n",
    "    6. Осталось собрать результаты каждого из вызовов — идём в обратную сторону: результат последнего вызова — 0, складываем его с 12, потом прибавляем 24, затем 21 и, наконец, 10. Получаем 67."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac613d2e",
   "metadata": {},
   "source": [
    "Вы могли обратить внимание на то, что при рекурсии выполнение функций, результат выполнения которых ещё не известен, «откладывается». Существует ли какое-то хранилище, в которое «откладываются» невыполненные функции? Да, и оно называется стэк вызова функций. Давайте познакомимся с этим понятием поближе."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6277ca7c",
   "metadata": {},
   "source": [
    "# СТЭК ВЫЗОВА ФУНКЦИЙ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bc1d0ca5",
   "metadata": {},
   "source": [
    "Представьте, что вы сидите за столом и проходите данный курс. В какой-то момент вы захотели сделать перерыв и пойти на кухню перекусить. Когда вы были на кухне, вам позвонил по телефону приятель, с которым вы начали беседовать. Вдруг в дверь позвонил курьер. Вы попросили друга повисеть на линии и отложили телефон, чтобы открыть дверь и взять вещи у курьера. \n",
    "\n",
    "Давайте остановимся на моменте открытия входной двери и посчитаем, сколько у вас сейчас прерванных задач, а также поймём, в каком порядке вы будете их завершать. Итак, вам необходимо:\n",
    "\n",
    "1. Взять сумки у курьера.\n",
    "2. Завершить разговор с другом.\n",
    "3. Закончить перекус.\n",
    "4. Вернуться к прохождению курса.\n",
    "\n",
    "Задачи перечислены в том порядке, в котором они и будут выполняться. Обратите внимание, что изначально эти задачи возникали в обратном порядке.\n",
    "\n",
    "А вот ещё одна важная аналогия: представьте, что вы моете посуду и складываете чистые тарелки друг на друга. В каком порядке вы будете брать тарелки, чтобы поставить их на полку для посуды? Правильно, в обратном — последняя помытая тарелка пойдёт на полку первой.\n",
    "\n",
    "Такая структура данных, в которой первым обрабатывается элемент (в данном случае задача), появившийся последним, называется **стэком** (или «рюкзаком»). Мы получили стэк задач. Более формально он и называется **стэком вызовов функции (call stack)**. В стэке действует правило **LIFO (last in — first out)** — «первый вошёл — последний вышел»."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "13e08de8",
   "metadata": {},
   "source": [
    "Когда Python при выполнении одной функции встречает шаг, на котором требуется выполнить другую функцию, он кладёт начальную функцию в стэк, пока выполняет вторую. Если и во второй окажется ссылка на другую функцию, вторая задача также будет отложена в стэк. Так будет продолжаться до тех пор, пока наконец какая-нибудь функция не вернёт конкретное значение. Тогда интерпретатор начнёт свой обратный путь по стэку для выполнения отложенных, ожидающих решения задач. Это мы и увидели в нашем примере."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "469744f4",
   "metadata": {},
   "source": [
    "На нашем простом примере, который на самом деле можно было решить в одну строчку кода (print(sum(lst))) мы рассмотрели алгоритм работы рекурсивных функций."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6094de9b",
   "metadata": {},
   "source": [
    "### Задание 3.4\n",
    "\n",
    "Напишите рекурсивную функцию multiply_lst(), которая перемножает элементы заданного списка между собой.\n",
    "\n",
    "Переведите алгоритм создания рекурсивной функции для расчёта суммы в расчёт произведения. Условие остановки — if len(lst) == 0: return 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "a5e3d725",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[10, 21, 24, 12]\n",
      "[21, 24, 12]\n",
      "[24, 12]\n",
      "[12]\n",
      "[]\n",
      "60480\n"
     ]
    }
   ],
   "source": [
    "def multiply_lst(lst):\n",
    "    print (lst)\n",
    "    if len(lst) == 0: return 1\n",
    "    return lst[0] * multiply_lst(lst[1:])\n",
    "\n",
    "my_lst = [10, 21, 24, 12]\n",
    "print(multiply_lst(my_lst))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c5b8e0e4",
   "metadata": {},
   "source": [
    "# ФАКТОРИАЛ ЧЕРЕЗ РЕКУРСИЮ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "34c591c4",
   "metadata": {},
   "source": [
    "Факториал возникает из решения следующей задачи: сколькими способами можно расставить n людей в очереди из n человек?\n",
    "\n",
    "Логично, что на первое место в очереди мы можем поставить любого из n человек, на второе — из n-1 человек, на третье — n-2 и т. д. На последнее место мы сможем поставить единственного оставшегося человека. Всего получается n*(n-1)* (n-2)*...* 2 * 1 возможных перестановок.\n",
    "\n",
    "Пусть, например, n=5. Тогда в очереди длиной в пять мест пять человек можно расставить 5 * 4 * 3 * 2 * 1= 120 различными способами.\n",
    "\n",
    "Эту запись в математике сокращают с помощью **функции факториал**, которая обозначается восклицательным знаком: n! (читается «эн факториал»). Итак, n!=n*(n-1)* (n-2) *...* 2 * 1.\n",
    "\n",
    "Например, 4! = 4 * 3 * 2 * 1 = 24\n",
    "Математики дополнительно договорились, что 0! = 1 и 1! = 1.\n",
    "\n",
    "Обратите внимание, что n! = n * (n-1) * (n-2) * ... * 2 * 1 = n * (n-1)! \n",
    "\n",
    "Из последней записи делаем вывод, что факториал можно записать с помощью рекурсивной функции: будем умножать n на (n-1)! до тех пор, пока под факториалом не окажется 0 или 1, для которых нам известно точное значение факториала."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "35120453",
   "metadata": {},
   "source": [
    "Итак, создадим рекурсивную функция factorial для вычисления факториала. Её аргументом будет целое число n. Функция будет рекурсивной, а условий остановки в ней будет сразу два: если n == 0 или n == 1 (в таком случае функция возвращает 1, об этом мы говорили ранее). Если условия остановки не выполняются, то функция возвращает результат умножения числа n на результат своей же работы для числа n-1: n * (n-1)!."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "9121f06a",
   "metadata": {},
   "outputs": [],
   "source": [
    "def factorial(n):\n",
    "    #Задаем условия выхода из рекурсии:\n",
    "    if n == 0: return 1\n",
    "    if n == 1: return 1\n",
    "    #во всех других случаях возвращаем произведение текущего числа n и функции\n",
    "    #От n - 1\n",
    "    return n * factorial(n-1)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "59e5f63c",
   "metadata": {},
   "source": [
    "Отлично, мы объявили нашу рекурсивную функцию. Давайте её протестируем — посчитаем факториал от 0, от 3 и от 5:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0ced949c",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "6\n",
      "120\n"
     ]
    }
   ],
   "source": [
    "print(factorial(0))\n",
    "print(factorial(3))\n",
    "print(factorial(5))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5e393eed",
   "metadata": {},
   "source": [
    "    1. Вызывается функция factorial(5). Стэк вызова пополняется вызовом первой функции. Локальная переменная n = 5. Условие выхода из рекурсии не выполняется (n != 1 и n!= 0), а значит выполняется основной блок функции. В этом блоке мы пытаемся вычислить выражение n * factorial(n-1), что равносильно 5 * factorial(4). Так как результат работы функции factorial(4) заранее не известен, то для начала мы должны её выполнить.\n",
    "\n",
    "\n",
    "    2. Вызывается функция factorial(4). Стэк вызова пополняется вызовом второй функции. Локальная переменная n = 4. Условие выхода всё ещё не выполняется, а значит выполняется основной блок функции. Снова пытаемся вычислить выражение n * factorial(n-1), что равносильно 4 * factorial(3). Но, чтобы получить результат, сначала должна выполниться функция factorial(3).\n",
    "\n",
    "    3. Вызывается функция factorial(3). Теперь длина стэка вызова равна 3. Всё повторяется вновь, только теперь n = 3. Вычисляем n  * factorial(n-1) = 3 * factorial(2). Переходим к вызову функции factorial(2).\n",
    "\n",
    "    4. Вызываем функцию factorial(2). Длина стэка вызова увеличивается до 4. Для выполнения функции нам необходимо вычислить n * factorial(n-1)= 2 * factorial(1). Мы всё ещё не получили конкретного ответа, поэтому переходим к вызову функции factorial(1).\n",
    "\n",
    "    5. Финал! Вызываем функцию factorial(1). Длина стэка вызова становится равной 5. В теле функции наконец выполняется условие остановки, так как n==1. Это значит, что factorial(1) возвращает конкретное значение 1.\n",
    "\n",
    "    6. А теперь обратное движение — освобождаем стэк вызова в противоположном его наполнению направлении. Последовательность завершения работы функций будет следующей:\n",
    "\n",
    "        1. На первом месте в стэке вызовов находится factorial(1) = 1. Функция отработала, поэтому стэк вызовов она покидает, его длина становится равной 4.\n",
    "\n",
    "        2. Затем выполняется функция factorial(2). Она должна вернуть 2 * factorial(1) = 2. После её выполнения стэк вызова снова уменьшается, и количество ожидающих задач становится равным 3.\n",
    "\n",
    "        3. Следующая на очереди — функция factorial(3). Она возвращает значение выражения 3 * factorial(2) = 3 * 2 = 6. После выполнения она покидает стэк. Длина стэка вызова составляет 2.\n",
    "        \n",
    "        4.Затем переходим к выполнению функции factorial(4). Её возвращаемое значение — результат выражения 4 * factorial(3) = 4 * 6 = 24. Функция отработала и покинула стэк. Его длина уменьшается до 1.\n",
    "        \n",
    "        5. Наконец, мы добрались до самой интересной для нас функции — factorial(5). Она должна вернуть результат выражения 5 * factorial(4) = 5 * 24 = 120. Процедура завершается, так как последняя функция покидает стэк вызова и вызывать больше нечего. Выполняются следующие строчки кода.\n",
    "\n",
    "\n",
    "Именно по такому принципу работают все рекурсивные функции."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "2ac378df",
   "metadata": {},
   "source": [
    "# ГЛУБИНА РЕКУРСИИ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8e7de821",
   "metadata": {},
   "source": [
    "Теперь поближе посмотрим на условия остановки, на две очень важные строки в коде нашей функции:\n",
    "\n",
    "#Задаём условия выхода из рекурсии:\n",
    "    if n==0: return 1\n",
    "    if n==1: return 1\n",
    "\n",
    "Они задают условие, при выполнении которого функция возвращает конкретное значение, а не продолжает ссылаться на саму себя. \n",
    "\n",
    "Давайте удалим условие выхода:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "d44acc60",
   "metadata": {},
   "outputs": [],
   "source": [
    "def wrong_factorial(n):\n",
    "    return wrong_factorial(n-1)*n\n",
    "wrong_factorial(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "7d7ddd98",
   "metadata": {},
   "source": [
    "Убило ядро.\n",
    "\n",
    "Мы знаем, что с рекурсией происходит наполнение стэка вызовов всё новыми и новыми вызовами рекурсивной функции. Очень тесно с этим явлением связан термин глубина рекурсии.\n",
    "\n",
    "Число единовременно ожидающих выполнения вызовов рекурсивной функции в стэке и называется глубиной рекурсии. Проще говоря, глубина рекурсии — это длина стэка вывода. \n",
    "\n",
    "Например, при расчёте факториала от 3 функция factorial будет вызвана три раза: для расчёта факториала от 3, 2 и 1 соответственно. В какой-то момент все три вызова окажутся в стэке вызовов, а глубина рекурсии составит 3.\n",
    "\n",
    "Когда глубина рекурсии оказывается слишком большой (превышает максимальный размер стэка функций), возникает ошибка RecursionError.\n",
    "\n",
    "→ Кстати, обратите внимание, что в представленном выше объяснении также присутствует вложенность, характерная для стэка вызовов: мы столкнулись с ошибкой RecursionError, для объяснения которой потребовалось понятие глубины рекурсии, которое в свою очередь потребовало введения термина стэка вызовов функции. После обсуждения стэка вызовов мы вернулись к глубине рекурсии и, наконец, уточнили причину появления ошибки RecursionError! Вот он, пример рекурсивных размышлений.\n",
    "\n",
    "На самом деле ошибка RecursionError может возникнуть не только для рекурсивной функции, для которой забыли написать условие выхода из рекурсии. К сожалению, интерпретатор не может понять, есть ли условие выхода в коде функции, поэтому, когда в его стэке накопилось слишком много вызовов той же самой функции, он вернёт ошибку (это защита от бесконечного выполнения программы). Этот порог индивидуален и может зависеть от сложности самой функции, версии Python и других настроек. \n",
    "\n",
    "На самом деле глубину рекурсии можно увеличить самостоятельно, если для решения вашей задачи крайне необходимо использовать рекурсию большей глубины. Вспомним первый пример — рекурсивный подсчёт суммы чисел в списке. Проблема состоит в том, что если бы длина списка была 1 000 элементов, то нам бы понадобилось 1 000 вызовов рекурсии, что, скорее всего, вызвало бы ошибку RecursionError. Тут-то нам и понадобится увеличение максимальной глубины рекурсии.\n",
    "\n",
    "Давайте увеличим максимально допустимую глубину рекурсии, например до 1 000 000 000. Для этого нам необходимо импортировать модуль sys, в котором содержатся функции для управления вашей системой. Оттуда нам понадобится функция setrecursionlimit, в аргументы которой нужно передать желаемую максимальную глубину рекурсии.\n",
    "\n",
    "После нашей маленькой настройки снова попробуем посчитать факториал от 970."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b44ffa54",
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'factorial' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[1;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "Input \u001b[1;32mIn [3]\u001b[0m, in \u001b[0;36m<cell line: 5>\u001b[1;34m()\u001b[0m\n\u001b[0;32m      3\u001b[0m \u001b[38;5;66;03m#увеличим глубину рекурсии\u001b[39;00m\n\u001b[0;32m      4\u001b[0m sys\u001b[38;5;241m.\u001b[39msetrecursionlimit(\u001b[38;5;241m1000000000\u001b[39m)\n\u001b[1;32m----> 5\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[38;5;28mlen\u001b[39m(\u001b[38;5;28mstr\u001b[39m(\u001b[43mfactorial\u001b[49m(\u001b[38;5;241m970\u001b[39m))))\n",
      "\u001b[1;31mNameError\u001b[0m: name 'factorial' is not defined"
     ]
    }
   ],
   "source": [
    "#импортируем модуль для работы с системными переменными\n",
    "import sys\n",
    "#увеличим глубину рекурсии\n",
    "sys.setrecursionlimit(1000000000)\n",
    "print(len(str(factorial(970))))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c46b799c",
   "metadata": {},
   "source": [
    "ХЗ ПОЧЕМУ НЕ ВЫДАЕТ"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c4d02ba6",
   "metadata": {},
   "source": [
    "→ Вы научились создавать рекурсивные функции с условием выхода из рекурсии. В Data Science вы обязательно столкнётесь с ситуациями, когда применение рекурсии покажется вам наиболее простым методом для решения задачи. Тот же перебор вложенных папок, в которых у вас могут находиться файлы, создание последовательностей чисел (например, числа Фибоначчи, применяемые для  определения отрезков времени, через которое произойдёт то или иное событие, например изменение цены) или генерация распределений.\n",
    "\n",
    "К тому же, рекурсии лежат в основе множества алгоритмов, в том числе алгоритмов машинного обучения.\n",
    "\n",
    "Однако рекурсии работают медленнее, чем аналогичные по сути циклы. К тому же существует ряд случаев, когда преобразование рекурсии в цикл затруднительно и неэффективно, например, реализация алгоритма дерева решений, быстрое преобразование Фурье или алгоритм quicksort. Однако, когда это оказывается возможным, цикл работает быстрее.\n",
    "\n",
    "Продемонстрируем это с помощью функции time из модуля time. Она позволяет засечь время выполнения фрагмента кода (в нашем случае — вызова функции factorial).\n",
    "\n",
    "Будем 100 раз считать факториал из 10000:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "84e5c56e",
   "metadata": {},
   "outputs": [],
   "source": [
    "from time import time\n",
    "import sys\n",
    "sys.setrecursionlimit(1000000000)\n",
    "\n",
    "def factorial(n):\n",
    "    if n==0: return 1\n",
    "    if n==1: return 1\n",
    "    return n * factorial(n-1)\n",
    "\n",
    "a = time()\n",
    "for i in range(100):\n",
    "    factorial(10000)\n",
    "b = time()\n",
    "print(a-b)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "6a648381",
   "metadata": {},
   "source": [
    "УБИЛО ЯДРО"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ac97fde6",
   "metadata": {},
   "source": [
    "ответ был: 4.058465242385864\n",
    "\n",
    "Чтобы 100 раз посчитать факториал от 10 000, потребовалось 4,06 секунды (при каждом вызове время может незначительно различаться)."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e65c25ec",
   "metadata": {},
   "source": [
    "А сколько времени потребуется для вычислений в цикле?\n",
    "\n",
    "Напишем функцию factorial_for, которая считает факториал числа n в цикле:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "38febe2d",
   "metadata": {},
   "outputs": [],
   "source": [
    "def factorial_for(n):\n",
    "    #для расчета произведения первый член - 1, а не 0!\n",
    "    result = 1\n",
    "    #перемножаем числа от 1 до n\n",
    "    for i in range (1, n+1):\n",
    "        result *= i\n",
    "    return result"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b1b140a4",
   "metadata": {},
   "source": [
    "Также посчитаем 100 раз факториал из 10 000:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "dcdb71e3",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "2.0065815448760986\n"
     ]
    }
   ],
   "source": [
    "from time import time\n",
    "# Объявляем функцию для расчёта факториала через цикл\n",
    "def factorial_for(n):\n",
    "    # Для расчёта произведения первый член — единица, а не ноль!\n",
    "    result = 1\n",
    "    # Перемножаем числа от 1 до n\n",
    "    for i in range(1, n+1):\n",
    "        result *= i\n",
    "    return result\n",
    "# Засекаем время до начала выполнения цикла\n",
    "a = time()\n",
    "for i in range(100):\n",
    "    factorial_for(10000)\n",
    "# Засекаем время после выполнения цикла\n",
    "b = time()\n",
    "# Считаем разницу: вычисляем время, потраченное на выполнение\n",
    "print(b-a)\n",
    "# Будет напечатано:\n",
    "# 3.0138471126556396"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fb4c7ddc",
   "metadata": {},
   "source": [
    "ЛОЛ.\n",
    "\n",
    "Потребовалось 3,01 секунды на расчёты, то есть почти на секунду (или почти на 25 %) меньше времени относительно рекурсии.\n",
    "\n",
    "→ Кто-то скажет, что одна секунда — это очень мало и не имеет значения. В рамках нашей задачи так оно и есть. Однако рекурсивные функции могут быть намного объёмнее по вычислениям, чем наш факториал: например, рекурсивная функция может два и более раз вызывать в своём теле саму себя. Если рекурсивная функция является частью большой программы жизнеобеспечения, где она вызывается сотни или тысячи раз, то её задержка по сравнению с циклом может составить гораздо больший промежуток, чем одна секунда."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ffe28f2e",
   "metadata": {},
   "source": [
    "### Задание 3.7\n",
    "\n",
    "Числа Фибоначчи — пример последовательности, которую можно получить рекурсивно. Вот первые числа этой последовательности: 0, 1, 1, 2, 3, 5, 8, 13 …\n",
    "\n",
    "Как уже отмечалось, последовательность Фибоначчи имеет важное значение в комбинаторике и аналитике.\n",
    "\n",
    "Как вы могли заметить, каждый -ый элемент — это сумма n-1 и n-2 элементов.\n",
    "\n",
    "Можно записать рекурсивное выражение для расчёта -ого числа из последовательности Фибоначчи:\n",
    "\n",
    "an = an-1 + an-2\n",
    "\n",
    "Напишите рекурсивную функцию fib(n), которая считает -ое число Фибоначчи. Будем считать, что fib(0) = 0 и fib(1) = 1."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "2a5606c0",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "1\n",
      "8\n"
     ]
    }
   ],
   "source": [
    "def fib(n):\n",
    "    if n == 0: return 0\n",
    "    if n == 1: return 1\n",
    "    return fib(n-1) + fib(n-2)\n",
    "\n",
    "print(fib(0))\n",
    "print(fib(2))\n",
    "print(fib(6))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "5301d3fc",
   "metadata": {},
   "source": [
    "### Задание 4.5\n",
    "\n",
    "Напишите рекурсивную функцию power(val, n), которая возводит число в заданную целую натуральную степень (или в степень 0).\n",
    "\n",
    "Использовать встроенный оператор ** для возведения в степень запрещено. Пользуйтесь только умножением *. Например, 2 ** 4 = (((2 * 2) * 2) * 2) = 16.\n",
    "\n",
    "В качестве первого аргумента функция принимает число, в качестве второго — желаемую степень.\n",
    "\n",
    "Примечание: в качестве базовых случаев для рекурсии возьмите условие, что любое число в степени 0 — 1, любое число в степени 1 — это же число."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "9bc23222",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "625\n"
     ]
    }
   ],
   "source": [
    "def power(val, n):\n",
    "    if n == 0: return 1\n",
    "    if n == 1: return val\n",
    "    return val * val ** (n-1)\n",
    "\n",
    "print(power(25,0))\n",
    "print(power(-5,4))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "8da4d146",
   "metadata": {},
   "source": [
    "### Задание 4.6\n",
    "\n",
    "Напишите функцию is_leap(year), которая принимает на вход год и возвращает True, если год високосный, иначе — False.\n",
    "\n",
    "Условие для проверки на високосность: год делится на 400 или год делится на 4, но не на 100.\n",
    "\n",
    "В теле функции задайте три условия: год нацело делится на 400, 100 и 4. Если выполняется первое и последнее условие, функция возвращает True, если не выполняется второе условие или не выполняется ни одно из перечисленных условий, функция возвращает False."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "19ad7fe9",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "False\n",
      "True\n"
     ]
    }
   ],
   "source": [
    "def is_leap(year):\n",
    "    if year % 400 == 0 or year % 4 == 0 and year % 100 != 0: \n",
    "        return True\n",
    "    else:\n",
    "        return False\n",
    "\n",
    "print(is_leap(2000))\n",
    "print(is_leap(1900))\n",
    "print(is_leap(2020))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b90b8110",
   "metadata": {},
   "source": [
    "### Задание 4.7\n",
    "\n",
    "При разработке пользовательского интерфейса и программ, взаимодействующих с человеком, очень важна проверка корректности введённых пользователем данных. Такую проверку в сообществе программистов принято называть «проверкой на дурака». Некорректно введённые данные могут «сломать» вашу программу, если она не была подготовлена для их обработки.\n",
    "\n",
    "Напишите функцию check_date(day, month, year), которая проверяет корректность даты рождения по следующим условиям:\n",
    "\n",
    "1. Все аргументы должны быть целыми числами (проверить с помощью type(day) is int).\n",
    "2. Годом рождения не может быть год до 1900 и год после 2022.\n",
    "3. Номер месяца не может быть больше 12 и меньше 1.\n",
    "4. Номер дня не может быть больше 31 и меньше 1.\n",
    "5. В сентябре, апреле, июне и ноябре 30 дней.\n",
    "6. Если год является високосным, то в феврале (второймесяц) должно быть 29 дней, в противном случае — 28.\n",
    "\n",
    "Если дата корректна, вернуть True, если же хотя бы одно из представленных условий не было выполнено — False. \n",
    "\n",
    "Примечание: использовать встроенные модули для работы с датами нельзя.\n",
    "\n",
    "Функция check_date должна содержать в себе функцию is_leap из предыдущего задания. В теле внешней функции вызывайте внутреннюю, чтобы проверить, является ли год високосным."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "53ae1f14",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "True\n",
      "True\n",
      "False\n",
      "False\n",
      "False\n"
     ]
    }
   ],
   "source": [
    "def check_date(day, month, year):\n",
    "    def is_leap(year):\n",
    "        if year % 400 == 0 or year % 4 == 0 and year % 100 != 0: \n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    if (type(day) is not int) or (type(month) is not int) or (type(year) is not int): \n",
    "        return False\n",
    "    if (year <= 1900) or (year >= 2022): \n",
    "        return False\n",
    "    if (month > 12) or (month < 1):\n",
    "        return False\n",
    "    if (day < 1) or (day > 31):\n",
    "        return False\n",
    "    if (month in [4,6,9,11]) and (day >30):\n",
    "        return False\n",
    "    if month == 2:\n",
    "        if is_leap(year):\n",
    "            if day > 29: return False\n",
    "        else:\n",
    "            if day > 28: return False\n",
    "    return True\n",
    "                \n",
    "        \n",
    "    \n",
    "print(check_date(18,9,1999))\n",
    "print(check_date(29,2,2000))\n",
    "print(check_date(29,2,2021))\n",
    "print(check_date(13,13,2021))\n",
    "print(check_date(13.5,12,2021))"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e626a608",
   "metadata": {},
   "source": [
    "### Задание 4.8\n",
    "\n",
    "Представьте, что вы пишете приложение, предусматривающее регистрацию пользователей. Давайте реализуем небольшой функционал регистрации. Не забудем также про «проверку на дурака».\n",
    "\n",
    "Напишите функцию register(surname, name, date, middle_name, registry).\n",
    "\n",
    "Функция имеет следующие аргументы:\n",
    "\n",
    "1. surname — фамилия;\n",
    "2. name — имя;\n",
    "2. date — дата рождения (в виде кортежа из трёх чисел — день, месяц, год);\n",
    "3. middle_name — отчество ;\n",
    "4. registry — список, в который необходимо добавить полученные аргументы в виде кортежа в следующем порядке: фамилия, имя, отчество, день, месяц, год рождения.\n",
    "\n",
    "Функция должна возвращать список, в который добавила запись. \n",
    "\n",
    "Указание: сделайте отчество аргументом по умолчанию со значением None, так как отчество может быть не у всех регистрирующихся.\n",
    "\n",
    "Также сделайте так, чтобы пустой список создавался в том случае, если он не был передан извне. Таким образом, по умолчанию registry имеет значение None, и если при вызове функции список не был передан, он создаётся в теле функции.\n",
    "\n",
    "Наконец, проверьте дату на корректность. Если дата неправильная, верните ошибку ValueError(\"Invalid Date!\"). Для этого вам пригодится функция check_date из предыдущего задания."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "0c6191fa",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[('Petrova', 'Maria', 'Ivanovna', 13, 3, 2003), ('Ivanov', 'Sergej', None, 24, 9, 1995), ('Smith', 'John', None, 13, 2, 2003)]\n"
     ]
    }
   ],
   "source": [
    "def check_date(day, month, year):\n",
    "    def is_leap(year):\n",
    "        if year % 400 == 0 or year % 4 == 0 and year % 100 != 0: \n",
    "            return True\n",
    "        else:\n",
    "            return False\n",
    "    if (type(day) is not int) or (type(month) is not int) or (type(year) is not int): \n",
    "        return False\n",
    "    if (year <= 1900) or (year >= 2022): \n",
    "        return False\n",
    "    if (month > 12) or (month < 1):\n",
    "        return False\n",
    "    if (day < 1) or (day > 31):\n",
    "        return False\n",
    "    if (month in [4,6,9,11]) and (day >30):\n",
    "        return False\n",
    "    if month == 2:\n",
    "        if is_leap(year):\n",
    "            if day > 29: return False\n",
    "        else:\n",
    "            if day > 28: return False\n",
    "    return True\n",
    "\n",
    "def register(surname, name, date, middle_name = None, registry = None):\n",
    "    if registry is None:\n",
    "        registry = list()\n",
    "    if not check_date(*date):\n",
    "        raise ValueError('Ivalid Date!')\n",
    "    registry.append((surname, name, middle_name, date[0], date[1], date[2]))\n",
    "    return registry\n",
    "\n",
    "reg = register('Petrova', 'Maria', (13, 3, 2003), 'Ivanovna')\n",
    "reg = register('Ivanov', 'Sergej', (24, 9, 1995), registry=reg)\n",
    "reg = register('Smith', 'John', (13, 2, 2003), registry=reg)\n",
    "print(reg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "569780bd",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
